### PR TITLE
refactor hyper parameters

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

(1) refactor hyper parameters of OptimMethod
(2) add OptimMethod object to python

## How was this patch tested?

manual test and unit test

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 1]
```diff
@@ -16,52 +16,78 @@
...
-    val nesterov = config.get[Boolean]("nesterov").getOrElse(false)
-    val lrs = config.get[Tensor[T]]("learningRates").getOrElse(null)
-    val wds = config.get[Tensor[T]]("weightDecays").getOrElse(null)
+    val wd = this.weightDecay
+    val mom = this.momentum
+    val damp = this.dampening
+    val nesterov = this.nesterov
+    val lrs = this.learningRates
+    val wds = this.weightDecays
+    val clr = ev.fromType(this.clr)
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: clr is useless

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 2]
```diff
@@ -189,8 +230,8 @@ object SGD {
    * @param maxIteration max iteration when lr becomes zero
    */
   case class Poly(power : Double, maxIteration : Int) extends LearningRateSchedule {
-    override def updateHyperParameter(config: Table, state: Table): Unit = {
-      val lr = config.get[Double]("learningRate").getOrElse(1e-3)
+    override def updateHyperParameter[T](optimMethod: SGD[T], state: Table): Unit = {
+      val lr = optimMethod.learningRate
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Can we change to a style `a(n) = a(n-1) ...`? So we abandon the config table.

**Comment by qiuxin2012** [comment]
Make sure the resume training works fine.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 2 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: no, and we have to keep original paramer value.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala (hunk 1 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: yeah, no clr in SGD now. clr is produced by leaningrate schedule. So clr is defined in leaningrate schedule.

File: pyspark/dl/optim/optimizer.py [hunk 3]
```diff
@@ -172,7 +346,8 @@ def __init__(self,
        :param batch_size: training batch size
        """ 
        JavaValue.__init__(self, None, bigdl_type, model.value,
-                           training_rdd, criterion, optim_method,
+                           training_rdd, criterion,
+                          optim_method if optim_method else SGD(),
```
Comment (zhichao-li) [inline pyspark/dl/optim/optimizer.py]: Add a doc to describe this default behavior?
optim_method = SGD() if optim_method is None else optim_method

File: pyspark/test/simple_integration_test.py [hunk 4]
```diff
@@ -73,7 +73,7 @@ def gen_rand_sample():
             model=model_test,
             training_rdd=trainingData,
             criterion=MSECriterion(),
-            optim_method="sgd",
+            optim_method=SGD(),
             state=state,
```
Comment (zhichao-li) [inline pyspark/test/simple_integration_test.py]: should we remove the state as well?

**Comment by zhichao-li** [comment]
Those examples under ```BigDL/pyspark/dl/models``` should be updated as well.

File: pyspark/test/simple_integration_test.py (hunk 4 above)
Comment (zhangxiaoli73) [inline pyspark/test/simple_integration_test.py]: yeah, state can be removed

File: pyspark/dl/optim/optimizer.py (hunk 3 above)
Comment (zhangxiaoli73) [inline pyspark/dl/optim/optimizer.py]: ok, added

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala [hunk 5]
```diff
@@ -68,15 +68,10 @@ object Train {
...
 
-      val state = if (param.stateSnapshot.isDefined) {
-        T.load(param.stateSnapshot.get)
+      val optimMethod = if (param.stateSnapshot.isDefined) {
+        OptimMethod.load[Float](param.stateSnapshot.get)
       } else {
-        T(
-          "learningRate" -> 0.01,
-          "weightDecay" -> 0.0005,
-          "momentum" -> 0.9,
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala]: There is no momentum in Adagrad. So delete it.

File: pyspark/dl/models/lenet/lenet5.py [hunk 6]
```diff
@@ -75,14 +75,11 @@ def get_minst(sc, data_type="train", location="/tmp/mnist"):
...
             normalizer(mnist.TEST_MEAN, mnist.TEST_STD))
-        state = {"learningRate": 0.01,
-                 "learningRateDecay": 0.0002}
         optimizer = Optimizer(
             model=build_model(10),
             training_rdd=train_data,
             criterion=ClassNLLCriterion(),
-            optim_method="SGD",
-            state=state,
+            optim_method=SGD(0.01, 0.0002),
```
Comment (yiheng) [inline pyspark/dl/models/lenet/lenet5.py]: SGD(learningRate=0.01, weightDecay=0.0002)

File: pyspark/dl/models/textclassifier/textclassifier.py [hunk 7]
```diff
@@ -132,17 +132,13 @@ def train(sc,
...
-
     optimizer = Optimizer(
         model=build_model(news20.CLASS_NUM),
         training_rdd=train_rdd,
         criterion=ClassNLLCriterion(),
         end_trigger=MaxEpoch(max_epoch),
         batch_size=batch_size,
-        optim_method="Adagrad",
-        state=state)
+        optim_method=Adagrad(0.01, 0.0002))
```
Comment (yiheng) [inline pyspark/dl/models/textclassifier/textclassifier.py]: Adagrad(learningRate=0.01, weightDecay=0.0002)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/utils/TextClassifier.scala [hunk 8]
```diff
@@ -235,10 +235,9 @@ class TextClassifier(param: AbstractTextClassificationParams) extends Serializab
       criterion = new ClassNLLCriterion[Float](),
       batchSize = param.batchSize
     )
-    val state = T("learningRate" -> 0.01, "learningRateDecay" -> 0.0002)
+
     optimizer
-      .setState(state)
-      .setOptimMethod(new Adagrad())
+      .setOptimMethod(new Adagrad(0.01, 0.0002))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/utils/TextClassifier.scala]: Adagrad(learningRate=..., weightDecay=...)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/example/utils/TextClassifier.scala [hunk 9]
```diff
@@ -274,10 +273,8 @@ class TextClassifier(param: AbstractTextClassificationParams) extends Serializab
       batchSize = param.batchSize
     )
 
-    val state = T("learningRate" -> 0.01, "learningRateDecay" -> 0.0002)
     optimizer
-      .setState(state)
-      .setOptimMethod(new Adagrad())
+      .setOptimMethod(new Adagrad(0.01, 0.0002))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/example/utils/TextClassifier.scala]: Adagrad(learningRate=..., weightDecay=...)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala [hunk 10]
```diff
@@ -68,15 +68,10 @@ object Train {
...
-          "dampening" -> 0.0
-        )
+        new Adagrad[Float](0.01, 0.0, 0.0005)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala]: Adagrad(learningRate=..., weightDecay=...)

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Train.scala [hunk 11]
```diff
@@ -63,24 +63,15 @@ object TrainInceptionV1 {
...
+        OptimMethod.load[Float](param.stateSnapshot.get)
       } else if (param.maxEpoch.isDefined) {
-        T(
-          "learningRate" -> param.learningRate,
-          "weightDecay" -> param.weightDecay,
-          "momentum" -> 0.9,
-          "dampening" -> 0.0,
-          "learingRateSchedule" -> SGD.Poly(0.5, math.ceil(1281167.toDouble / param.batchSize).toInt
-            * param.maxEpoch.get))
+        new SGD[Float](
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Train.scala]: add the name of the parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Train.scala [hunk 12]
```diff
@@ -63,24 +63,15 @@ object TrainInceptionV1 {
...
+          SGD.Poly(0.5, math.ceil(1281167.toDouble / param.batchSize).toInt * param.maxEpoch.get))
       } else {
-        T(
-          "learningRate" -> param.learningRate,
-          "weightDecay" -> param.weightDecay,
-          "momentum" -> 0.9,
-          "dampening" -> 0.0,
-          "learningRateSchedule" -> SGD.Poly(0.5, param.maxIteration)
-        )
+        new SGD[Float](param.learningRate, 0.0, param.weightDecay, 0.9, 0.0,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/inception/Train.scala]: add the name of the parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/Train.scala [hunk 13]
```diff
@@ -53,13 +51,10 @@ object Train {
...
-      val state = if (param.stateSnapshot.isDefined) {
-        T.load(param.stateSnapshot.get)
+      val optimMethod = if (param.stateSnapshot.isDefined) {
+        OptimMethod.load[Float](param.stateSnapshot.get)
       } else {
-        T(
-          "learningRate" -> param.learningRate,
-          "learningRateDecay" -> param.learningRateDecay
-        )
+        new SGD[Float](param.learningRate, param.learningRateDecay)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/lenet/Train.scala]: add the name of the parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Train.scala [hunk 14]
```diff
@@ -67,17 +67,11 @@ object Train {
...
       } else {
-        T(
-          "learningRate" -> param.learningRate,
-          "weightDecay" -> param.weightDecay,
-          "momentum" -> param.momentum,
-          "dampening" -> param.dampening,
-          "nesterov" -> param.nesterov,
-          "learningRateSchedule" -> SGD.EpochDecay(cifar10Decay)
-        )
+        new SGD[Float](param.learningRate, 0.0, param.weightDecay, param.momentum,
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Train.scala]: add the name of the parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala [hunk 15]
```diff
@@ -100,13 +100,10 @@ object Train {
...
-      val state = if (param.stateSnapshot.isDefined) {
-        T.load(param.stateSnapshot.get)
+      val optimMethod = if (param.stateSnapshot.isDefined) {
+        OptimMethod.load[Float](param.stateSnapshot.get)
       } else {
-        T("learningRate" -> param.learningRate,
-          "momentum" -> param.momentum,
-          "weightDecay" -> param.weightDecay,
-          "dampening" -> param.dampening)
+        new SGD[Float](param.learningRate, 0.0, param.weightDecay, param.momentum, param.dampening)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/rnn/Train.scala]: add the name of the parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/vgg/Train.scala [hunk 16]
```diff
@@ -47,16 +47,10 @@ object Train {
...
+        OptimMethod.load[Float](param.stateSnapshot.get)
       } else {
-        T(
-          "learningRate" -> 0.01,
-          "weightDecay" -> 0.0005,
-          "momentum" -> 0.9,
-          "dampening" -> 0.0,
-          "learningRateSchedule" -> SGD.EpochStep(25, 0.5)
-        )
+        new SGD[Float](0.01, 0.0, 0.0005, 0.9, 0.0, false, SGD.EpochStep(25, 0.5))
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/vgg/Train.scala]: add the name of the parameters

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala [hunk 17]
```diff
@@ -17,12 +17,91 @@
...
+    File.save(this, path, overWrite)
+    this
+  }
+
+  /**
+   * generate optimMethod parameters from Table
+   * @param config
+   * @return
+   */
+  def applyTable(config: Table): this.type
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala]: rename it to load()?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala [hunk 18]
```diff
@@ -17,12 +17,91 @@
...
+   * @return
+   */
+  def save(path: String, overWrite: Boolean = false): this.type = {
+    this.clearHistory()
+    File.save(this, path, overWrite)
+    this
+  }
+
+  /**
+   * generate optimMethod parameters from Table
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala]: load optimMethod parameters from Table

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala (hunk 17 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/OptimMethod.scala]: yeah, I rename it to loadFromTable

File: pyspark/dl/optim/optimizer.py [hunk 19]
```diff
@@ -140,6 +142,179 @@ class Step(JavaValue):
...
+    """
+    def __init__(self,
+                 max_iter=20,
+                 max_eval=DOUBLEMAX,
+                 tolfun=1e-5,
+                 tolx=1e-9,
+                 ncorrection=100,
+                 learningrate=1.0,
+                 verbose=False,
+                 linesearch=None,
```
Comment (shane-huang) [inline pyspark/dl/optim/optimizer.py]: how should a user specify a linesearch function? It is a scala object? Would it work if a user defines a python linesearch function? If we don't have support python function, we can still have the argument here but throw an exception saying we don't support customized python linesearch function yet.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala (hunk 5 above)
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala]: As far as I remember, Qiu xin just added momentum support into Adagrad. Need to check this.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Train.scala [hunk 20]
```diff
@@ -67,17 +67,12 @@ object Train {
         curModel
       }
 
-      val state = if (param.stateSnapshot.isDefined) {
-        T.load(param.stateSnapshot.get)
+      val optimMethod = if (param.stateSnapshot.isDefined) {
+        OptimMethod.load[Float](param.stateSnapshot.get)
```
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Train.scala]: I'm curious what parameters are stored in state snapshot now? Only the optimMethod hyper params? If there's (or will be) any options other than hyper parameters, would it be better if the state has a method like "getHyperParams"?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adagrad.scala [hunk 21]
```diff
@@ -25,36 +25,32 @@ import scala.reflect.ClassTag
...
   /**
    * Adagrad implementation for Adagrad
    *
    * @param feval     a function that takes a single input (X), the point of a evaluation, and
    *                  returns f(X) and df/dX
    * @param parameter the initial point
-   * @param config    a table with configuration parameters for the optimizer
-   *                  config("learningRate") : learning rate
-   *                  config("learningRateDecay") : learning rate decay
-   * @param state     a table describing the state of the optimizer; after each call the state
```
Comment (shane-huang) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adagrad.scala]: State is used to pass parameters to outside or used for snapshot, right? 
any documentations or rules which parameters are included in state?
Will it be a problem, if in training, a user uses one kind of optimMethod, e.g. SGD, and pauses, then resume the training using another kind of optimMethod, e.g. LBFGS?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Train.scala (hunk 20 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/Train.scala]: Now, the snapshot perserves the whole OptimMethod object, the object contains: necessary hyper parameters(learningRate...) and internal parameters(epoch, neval).

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adagrad.scala (hunk 21 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Adagrad.scala]: no documentations to describe state, by analyzing the code, the state includes intermediate params of optimMethod and internal parameters(epoch, neval).
Now, resume training from snapshot must have same optimMethod with training. I am not sure whether it is a problem.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala (hunk 5 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/autoencoder/Train.scala]: I check the code again, there is no momentum in Adagrad.

File: pyspark/dl/optim/optimizer.py (hunk 19 above)
Comment (zhangxiaoli73) [inline pyspark/dl/optim/optimizer.py]: linesearch function is a scala object, but no corresponding python function. Perhaps I can add one.

**Comment by zhangxiaoli73** [comment]
jenkins pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala [hunk 22]
```diff
@@ -201,6 +287,19 @@ object SGD {
...
+
+    override def updateHyperParameter[T](optimMethod: SGD[T]): Unit = {
+      val lr = optimMethod.learningRate
+      val nevals = optimMethod.state.get[Int]("evalCounter").getOrElse(0)
+      val clr = if (nevals > maxIteration) {
+        0.0
+      } else {
+        -lr * math.pow(1.0 - nevals.toDouble / maxIteration, power)
+      }
+      println(s"iteration is : ${nevals}. current learning rate is $clr")
```
Comment (qiuxin2012) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/SGD.scala]: Could you delete println?

**Pull request merged**